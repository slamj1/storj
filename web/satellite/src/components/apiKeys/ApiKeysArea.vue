// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <div class="api-keys-area">
        <h1 class="api-keys-area__title" v-if="!isEmpty">API Keys</h1>
        <div class="api-keys-area__container">
            <ApiKeysCreationPopup
                @closePopup="closeNewApiKeyPopup"
                @showCopyPopup="showCopyApiKeyPopup"
                :is-popup-shown="isNewApiKeyPopupShown"
            />
            <ApiKeysCopyPopup
                :is-popup-shown="isCopyApiKeyPopupShown"
                :api-key-secret="apiKeySecret"
                @closePopup="closeCopyNewApiKeyPopup"
            />
            <div v-if="isHeaderShown" class="api-keys-header">
                <VHeader
                    ref="headerComponent"
                    placeholder="API Key"
                    :search="onSearchQueryCallback">
                    <div class="header-default-state" v-if="isDefaultHeaderState">
                        <VButton
                            class="button"
                            label="+ Create API Key"
                            width="180px"
                            height="48px"
                            :on-press="onCreateApiKeyClick"
                        />
                    </div>
                    <div class="header-selected-api-keys" v-if="areApiKeysSelected">
                        <VButton
                            class="button deletion"
                            label="Delete"
                            width="122px"
                            height="48px"
                            :on-press="onFirstDeleteClick"
                        />
                        <VButton
                            class="button"
                            label="Cancel"
                            width="122px"
                            height="48px"
                            is-white="true"
                            :on-press="onClearSelection"
                        />
                        <span class="header-selected-api-keys__info-text"><b>{{selectedAPIKeysCount}}</b> API Keys selected</span>
                    </div>
                    <div class="header-after-delete-click" v-if="areSelectedApiKeysBeingDeleted">
                        <span class="header-after-delete-click__confirmation-label">Are you sure you want to delete <b>{{selectedAPIKeysCount}}</b> {{apiKeyCountTitle}} ?</span>
                        <div class="header-after-delete-click__button-area">
                            <VButton
                                class="button deletion"
                                label="Delete"
                                width="122px"
                                height="48px"
                                :on-press="onDelete"
                            />
                            <VButton
                                class="button"
                                label="Cancel"
                                width="122px"
                                height="48px"
                                is-white="true"
                                :on-press="onClearSelection"
                            />
                        </div>
                    </div>
                </VHeader>
                <div class="blur-content" v-if="isDeleteClicked"></div>
                <div class="blur-search" v-if="isDeleteClicked"></div>
            </div>
            <div v-if="!isEmpty" class="api-keys-items">
                <SortingHeader :on-header-click-callback="onHeaderSectionClickCallback"/>
                <div class="api-keys-items__content">
                    <VList
                        :data-set="apiKeyList"
                        :item-component="itemComponent"
                        :on-item-click="toggleSelection"
                    />
                </div>
                <VPagination
                    v-if="totalPageCount > 1"
                    class="pagination-area"
                    ref="pagination"
                    :total-page-count="totalPageCount"
                    :on-page-click-callback="onPageClick"
                />
            </div>
            <div class="empty-search-result-area" v-if="isEmptySearchResultShown">
                <h1 class="empty-search-result-area__title">No results found</h1>
                <EmptySearchResultIcon class="empty-search-result-area__image"/>
            </div>
            <NoApiKeysArea
                :on-button-click="onCreateApiKeyClick"
                v-if="isEmptyStateShown"
            />
        </div>
    </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';

import ApiKeysItem from '@/components/apiKeys/ApiKeysItem.vue';
import NoApiKeysArea from '@/components/apiKeys/NoApiKeysArea.vue';
import SortingHeader from '@/components/apiKeys/SortingHeader.vue';
import VButton from '@/components/common/VButton.vue';
import VHeader from '@/components/common/VHeader.vue';
import VList from '@/components/common/VList.vue';
import VPagination from '@/components/common/VPagination.vue';

import EmptySearchResultIcon from '@/../static/images/common/emptySearchResult.svg';

import { ApiKey, ApiKeyOrderBy } from '@/types/apiKeys';
import { SortDirection } from '@/types/common';
import { API_KEYS_ACTIONS } from '@/utils/constants/actionNames';
import { SegmentEvent } from '@/utils/constants/analyticsEventNames';

import ApiKeysCopyPopup from './ApiKeysCopyPopup.vue';
import ApiKeysCreationPopup from './ApiKeysCreationPopup.vue';

// header state depends on api key selection state
enum HeaderState {
    DEFAULT = 0,
    ON_SELECT,
}

const {
    FETCH,
    DELETE,
    TOGGLE_SELECTION,
    CLEAR,
    CLEAR_SELECTION,
    SET_SEARCH_QUERY,
    SET_SORT_BY,
    SET_SORT_DIRECTION,
} = API_KEYS_ACTIONS;

declare interface ResetPagination {
    resetPageIndex(): void;
}

@Component({
    components: {
        NoApiKeysArea,
        VList,
        VHeader,
        ApiKeysItem,
        VButton,
        ApiKeysCreationPopup,
        ApiKeysCopyPopup,
        VPagination,
        SortingHeader,
        EmptySearchResultIcon,
    },
})
export default class ApiKeysArea extends Vue {
    private FIRST_PAGE = 1;
    /**
     * Indicates if delete confirmation state should appear.
     */
    private isDeleteClicked: boolean = false;
    /**
     * Indicates if api key name input state should appear.
     */
    private isNewApiKeyPopupShown: boolean = false;
    /**
     * Indicates if copy api key state should appear.
     * Should only appear once
     */
    private isCopyApiKeyPopupShown: boolean = false;
    private apiKeySecret: string = '';

    public $refs!: {
        pagination: HTMLElement & ResetPagination;
    };

    /**
     * Lifecycle hook after initial render where list of existing api keys is fetched.
     */
    public async mounted(): Promise<void> {
        await this.$store.dispatch(FETCH, 1);
        this.$segment.track(SegmentEvent.API_KEYS_VIEWED, {
            project_id: this.$store.getters.selectedProject.id,
            api_keys_count: this.selectedAPIKeysCount,
        });
    }

    /**
     * Lifecycle hook before component destruction.
     * Clears existing api keys selection and search.
     */
    public async beforeDestroy(): Promise<void> {
        this.onClearSelection();
        await this.$store.dispatch(SET_SEARCH_QUERY, '');
    }

    /**
     * toggles api key selection.
     * @param apiKey
     */
    public async toggleSelection(apiKey: ApiKey): Promise<void> {
        await this.$store.dispatch(TOGGLE_SELECTION, apiKey);
    }

    public onCreateApiKeyClick(): void {
        this.isNewApiKeyPopupShown = true;
    }

    public onFirstDeleteClick(): void {
        this.isDeleteClicked = true;
    }

    public onClearSelection(): void {
        this.$store.dispatch(CLEAR_SELECTION);
        this.isDeleteClicked = false;
    }

    public closeNewApiKeyPopup(): void {
        this.isNewApiKeyPopupShown = false;
    }

    public showCopyApiKeyPopup(secret: string): void {
        this.isCopyApiKeyPopupShown = true;
        this.apiKeySecret = secret;
    }

    public closeCopyNewApiKeyPopup(): void {
        this.isCopyApiKeyPopupShown = false;
    }

    /**
     * Deletes selected api keys, fetches updated list and changes area state to default.
     */
    public async onDelete(): Promise<void> {
        try {
            await this.$store.dispatch(DELETE);
            await this.$notify.success(`API keys deleted successfully`);
            this.$segment.track(SegmentEvent.API_KEY_DELETED, {
                project_id: this.$store.getters.selectedProject.id,
            });
        } catch (error) {
            await this.$notify.error(error.message);
        }

        try {
            await this.$store.dispatch(FETCH, this.FIRST_PAGE);
        } catch (error) {
            await this.$notify.error(`Unable to fetch API keys. ${error.message}`);
        }

        this.isDeleteClicked = false;

        if (this.totalPageCount > 1) {
            this.$refs.pagination.resetPageIndex();
        }
    }

    public get itemComponent() {
        return ApiKeysItem;
    }

    /**
     * Returns api keys from store.
     */
    public get apiKeyList(): ApiKey[] {
        return this.$store.getters.apiKeys;
    }

    /**
     * Returns api keys pages count from store.
     */
    public get totalPageCount(): number {
        return this.$store.state.apiKeysModule.page.pageCount;
    }

    /**
     * Returns api keys label depends on api keys count.
     */
    public get apiKeyCountTitle(): string {
        return this.selectedAPIKeysCount === 1 ? 'api key' : 'api keys';
    }

    /**
     * Indicates if no api keys in store.
     */
    public get isEmpty(): boolean {
        return this.$store.getters.apiKeys.length === 0;
    }

    public get hasSearchQuery(): boolean {
        return this.$store.state.apiKeysModule.cursor.search;
    }

    public get selectedAPIKeysCount(): number {
        return this.$store.state.apiKeysModule.selectedApiKeysIds.length;
    }

    public get headerState(): number {
        return this.selectedAPIKeysCount > 0 ? HeaderState.ON_SELECT : HeaderState.DEFAULT;
    }

    public get isHeaderShown(): boolean {
        return !this.isEmpty || this.hasSearchQuery;
    }

    public get isDefaultHeaderState(): boolean {
        return this.headerState === HeaderState.DEFAULT;
    }

    public get areApiKeysSelected(): boolean {
        return this.headerState === HeaderState.ON_SELECT && !this.isDeleteClicked;
    }

    public get areSelectedApiKeysBeingDeleted(): boolean {
        return this.headerState === HeaderState.ON_SELECT && this.isDeleteClicked;
    }

    public get isEmptySearchResultShown(): boolean {
        return this.isEmpty && this.hasSearchQuery;
    }

    public get isEmptyStateShown(): boolean {
        return this.isEmpty && !this.isNewApiKeyPopupShown && !this.hasSearchQuery;
    }

    /**
     * Fetches api keys page by clicked index.
     * @param index
     */
    public async onPageClick(index: number): Promise<void> {
        try {
            await this.$store.dispatch(FETCH, index);
        } catch (error) {
            await this.$notify.error(`Unable to fetch API keys. ${error.message}`);
        }
    }

    /**
     * Used for sorting.
     * @param sortBy
     * @param sortDirection
     */
    public async onHeaderSectionClickCallback(sortBy: ApiKeyOrderBy, sortDirection: SortDirection): Promise<void> {
        await this.$store.dispatch(SET_SORT_BY, sortBy);
        await this.$store.dispatch(SET_SORT_DIRECTION, sortDirection);
        try {
            await this.$store.dispatch(FETCH, this.FIRST_PAGE);
        } catch (error) {
            await this.$notify.error(`Unable to fetch API keys. ${error.message}`);
        }

        if (this.totalPageCount > 1) {
            this.$refs.pagination.resetPageIndex();
        }
    }

    /**
     * Sets api keys search query and then fetches depends on it.
     * @param query
     */
    public async onSearchQueryCallback(query: string): Promise<void> {
        await this.$store.dispatch(SET_SEARCH_QUERY, query);
        try {
            await this.$store.dispatch(FETCH, this.FIRST_PAGE);
        } catch (error) {
            await this.$notify.error(`Unable to fetch API keys. ${error.message}`);
        }

        if (this.totalPageCount > 1) {
            this.$refs.pagination.resetPageIndex();
        }
    }
}
</script>

<style scoped lang="scss">
    .api-keys-area {
        position: relative;
        padding: 40px 65px 55px 65px;
        font-family: 'font_regular', sans-serif;

        &__title {
            font-family: 'font_bold', sans-serif;
            font-size: 32px;
            line-height: 39px;
            color: #263549;
            margin: 0;
        }

        .api-keys-header {
            width: 100%;
            position: relative;

            .blur-content {
                position: absolute;
                top: 100%;
                left: 0;
                background-color: #f5f6fa;
                width: 100%;
                height: 70vh;
                z-index: 100;
                opacity: 0.3;
            }

            .blur-search {
                position: absolute;
                bottom: 0;
                right: 0;
                width: 540px;
                height: 56px;
                z-index: 100;
                opacity: 0.3;
                background-color: #f5f6fa;
            }
        }

        .api-keys-items {
            position: relative;

            &__content {
                display: flex;
                flex-direction: column;
                width: 100%;
                justify-content: flex-start;
            }
        }
    }

    .empty-search-result-area {
        display: flex;
        align-items: center;
        justify-content: center;
        flex-direction: column;

        &__title {
            font-family: 'font_bold', sans-serif;
            font-size: 32px;
            line-height: 39px;
            margin-top: 104px;
        }

        &__image {
            margin-top: 40px;
        }
    }

    .pagination-area {
        margin-left: -25px;
        padding-bottom: 15px;
    }

    .header-default-state,
    .header-selected-api-keys {
        display: flex;
        align-items: center;

        &__info-text {
            margin-left: 25px;
            line-height: 48px;
        }
    }

    .header-selected-api-keys {

        .deletion {
            margin-right: 12px;
        }
    }

    .header-after-delete-click {
        display: flex;
        flex-direction: column;
        margin-top: 2px;

        &__confirmation-label {
            font-family: 'font_medium', sans-serif;
            font-size: 14px;
            line-height: 28px;
        }

        &__button-area {
            display: flex;
            margin-top: 4px;

            .button {
                margin-top: 2px;
            }

            .deletion {
                margin: 3px 12px 0 0;
            }
        }
    }

    .container.deletion {
        background-color: #ff4f4d;

        &.label {
            color: #fff;
        }

        &:hover {
            background-color: #de3e3d;
            box-shadow: none;
        }
    }

    .collapsed {
        margin-top: 0 !important;
        padding-top: 0 !important;
    }

    ::-webkit-scrollbar,
    ::-webkit-scrollbar-track,
    ::-webkit-scrollbar-thumb {
        width: 0;
    }

    @media screen and (max-width: 1024px) {

        .api-keys-area {
            padding: 40px 40px 55px 40px;
        }
    }
</style>
