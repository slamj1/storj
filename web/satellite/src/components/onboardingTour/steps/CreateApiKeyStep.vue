// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <div class="create-api-key-step">
        <h1 class="create-api-key-step__title">Create an API Key</h1>
        <p class="create-api-key-step__sub-title">
            API keys provide access to the project for creating buckets and uploading objects through the command line
            interface. This will be your first API key, and you can always create more keys later on.
        </p>
        <div class="create-api-key-step__container">
            <div class="create-api-key-step__container__title-area">
                <h2 class="create-api-key-step__container__title-area__title">Create API Key</h2>
                <img
                    v-if="isLoading"
                    class="create-api-key-step__container__title-area__loading-image"
                    src="@/../static/images/account/billing/loading.gif"
                    alt="loading gif"
                >
            </div>
            <HeaderedInput
                label="API Key Name"
                placeholder="Enter API Key Name (i.e. Dan’s Key)"
                class="full-input"
                width="calc(100% - 4px)"
                :error="errorMessage"
                @setData="setApiKeyName"
            />
            <div class="create-api-key-step__container__create-key-area" v-if="isCreatingState">
                <VButton
                    class="generate-button"
                    width="100%"
                    height="40px"
                    label="Generate API Key"
                    :is-blue-white="true"
                    :on-press="createApiKey"
                />
            </div>
            <div class="create-api-key-step__container__copy-key-area" v-else>
                <div class="create-api-key-step__container__copy-key-area__header">
                    <InfoImage/>
                    <span class="create-api-key-step__container__copy-key-area__header__title">
                        API Keys only appear here once. Copy and paste this key to your preferred method of storing secrets.
                    </span>
                </div>
                <div class="create-api-key-step__container__copy-key-area__key-container">
                    <span class="create-api-key-step__container__copy-key-area__key-container__key">{{ key }}</span>
                    <div class="create-api-key-step__container__copy-key-area__key-container__copy-button">
                        <VButton
                            width="81px"
                            height="40px"
                            label="Copy"
                            :is-blue-white="true"
                            :on-press="onCopyClick"
                        />
                    </div>
                </div>
            </div>
            <p class="create-api-key-step__container__info" v-if="isCopyState">
                We don’t record your API Keys, which are only displayed once when generated. If you loose this
                key, it cannot be recovered – but you can always create new API Keys when needed.
            </p>
            <div class="create-api-key-step__container__blur" v-if="isLoading"/>
        </div>
        <VButton
            class="done-button"
            width="156px"
            height="48px"
            label="Done"
            :on-press="onDoneClick"
            :is-disabled="isCreatingState"
        />
    </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';

import HeaderedInput from '@/components/common/HeaderedInput.vue';
import VButton from '@/components/common/VButton.vue';

import InfoImage from '@/../static/images/onboardingTour/info.svg';

import { ApiKey } from '@/types/apiKeys';
import { API_KEYS_ACTIONS } from '@/utils/constants/actionNames';
import { SegmentEvent } from '@/utils/constants/analyticsEventNames';
import { AddingApiKeyState } from '@/utils/constants/onboardingTourEnums';

const {
    CREATE,
    FETCH,
} = API_KEYS_ACTIONS;

@Component({
    components: {
        VButton,
        HeaderedInput,
        InfoImage,
    },
})
export default class CreateApiKeyStep extends Vue {
    private name: string = '';
    private addingState: number = AddingApiKeyState.CREATE;
    private readonly FIRST_PAGE = 1;

    public key: string = '';
    public errorMessage: string = '';
    public isLoading: boolean = false;

    /**
     * Indicates if view is in creating state.
     */
    public get isCreatingState(): boolean {
        return this.addingState === AddingApiKeyState.CREATE;
    }

    /**
     * Indicates if view is in copy state.
     */
    public get isCopyState(): boolean {
        return this.addingState === AddingApiKeyState.COPY;
    }

    /**
     * Indicates view state to copy state.
     */
    public setCopyState(): void {
        this.addingState = AddingApiKeyState.COPY;
    }

    /**
     * Sets api key name from input value.
     */
    public setApiKeyName(value: string): void {
        this.name = value.trim();
        this.errorMessage = '';
    }

    /**
     * Creates api key and refreshes store.
     */
    public async createApiKey(): Promise<void> {
        if (this.isLoading) {
            return;
        }

        if (!this.name) {
            this.errorMessage = 'API Key name can`t be empty';

            return;
        }

        this.isLoading = true;

        let createdApiKey: ApiKey;

        try {
            createdApiKey = await this.$store.dispatch(CREATE, this.name);
        } catch (error) {
            await this.$notify.error(error.message);
            this.isLoading = false;

            return;
        }

        await this.$notify.success('Successfully created new api key');
        this.key = createdApiKey.secret;

        this.$segment.track(SegmentEvent.API_KEY_CREATED, {
            project_id: this.$store.getters.selectedProject.id,
        });

        try {
            await this.$store.dispatch(FETCH, this.FIRST_PAGE);
        } catch (error) {
            await this.$notify.error(`Unable to fetch API keys. ${error.message}`);
        }

        this.setCopyState();

        this.isLoading = false;
    }

    /**
     * Copies api key secret to buffer.
     */
    public onCopyClick(): void {
        this.$copyText(this.key);
        this.$notify.success('Key saved to clipboard');
    }

    /**
     * Sets tour state to last step.
     */
    public onDoneClick(): void {
        this.$emit('setUploadDataState');
    }
}
</script>

<style scoped lang="scss">
    h1,
    h2,
    p {
        margin: 0;
    }

    .create-api-key-step {
        font-family: 'font_regular', sans-serif;
        margin-top: 75px;
        display: flex;
        flex-direction: column;
        align-items: center;
        justify-content: space-between;
        padding: 0 200px;

        &__title {
            font-size: 32px;
            line-height: 39px;
            color: #1b2533;
            margin-bottom: 25px;
        }

        &__sub-title {
            font-size: 16px;
            line-height: 19px;
            color: #354049;
            margin-bottom: 35px;
            text-align: center;
            word-break: break-word;
        }

        &__container {
            padding: 50px;
            width: calc(100% - 100px);
            border-radius: 8px;
            background-color: #fff;
            position: relative;
            margin-bottom: 30px;

            &__title-area {
                display: flex;
                align-items: center;
                justify-content: flex-start;
                margin-bottom: 10px;

                &__title {
                    font-family: 'font_medium', sans-serif;
                    font-size: 22px;
                    line-height: 27px;
                    color: #354049;
                    margin-right: 15px;
                }

                &__loading-image {
                    width: 18px;
                    height: 18px;
                }
            }

            &__create-key-area {
                width: calc(100% - 90px);
                padding: 40px 45px;
                margin-top: 30px;
                background-color: #0c2546;
                border-radius: 8px;
            }

            &__copy-key-area {
                width: 100%;
                margin-top: 30px;

                &__header {
                    padding: 10px;
                    width: calc(100% - 20px);
                    background-color: #ce3030;
                    display: flex;
                    align-items: center;
                    justify-content: flex-start;
                    border-radius: 8px 8px 0 0;

                    &__title {
                        font-size: 12px;
                        line-height: 16px;
                        color: #fff;
                    }
                }

                &__key-container {
                    background-color: #0c2546;
                    display: flex;
                    align-items: center;
                    padding: 20px 25px;
                    width: calc(100% - 50px);
                    border-radius: 0 0 8px 8px;

                    &__key {
                        font-size: 15px;
                        line-height: 23px;
                        color: #fff;
                        margin-right: 50px;
                        word-break: break-all;
                    }

                    &__copy-button {
                        min-width: 85px;
                    }
                }
            }

            &__info {
                width: 100%;
                margin-top: 30px;
                font-size: 12px;
                line-height: 18px;
                text-align: center;
                color: #354049;
                word-break: break-word;
            }

            &__blur {
                position: absolute;
                top: 0;
                left: 0;
                height: 100%;
                width: 100%;
                background-color: rgba(229, 229, 229, 0.2);
                z-index: 100;
            }
        }
    }

    .full-input {
        width: 100%;
    }

    .info-svg {
        min-width: 18px;
        margin-right: 5px;
    }

    @media screen and (max-width: 1650px) {

        .create-api-key-step {
            padding: 0 100px;
        }
    }

    @media screen and (max-width: 1450px) {

        .create-api-key-step {
            padding: 0 60px;
        }
    }

    @media screen and (max-width: 900px) {

        .create-api-key-step {
            padding: 0 50px;
        }
    }
</style>
