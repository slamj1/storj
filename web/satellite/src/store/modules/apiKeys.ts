// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { StoreModule } from '@/store';
import { ApiKey, ApiKeyCursor, ApiKeyOrderBy, ApiKeysApi, ApiKeysPage } from '@/types/apiKeys';
import { SortDirection } from '@/types/common';

export const API_KEYS_MUTATIONS = {
    SET_PAGE: 'setApiKeys',
    TOGGLE_SELECTION: 'toggleApiKeysSelection',
    CLEAR_SELECTION: 'clearApiKeysSelection',
    CLEAR: 'clearApiKeys',
    CHANGE_SORT_ORDER: 'changeApiKeysSortOrder',
    CHANGE_SORT_ORDER_DIRECTION: 'changeApiKeysSortOrderDirection',
    SET_SEARCH_QUERY: 'setApiKeysSearchQuery',
    SET_PAGE_NUMBER: 'setApiKeysPage',
};

const {
    SET_PAGE,
    TOGGLE_SELECTION,
    CLEAR_SELECTION,
    CLEAR,
    CHANGE_SORT_ORDER,
    CHANGE_SORT_ORDER_DIRECTION,
    SET_SEARCH_QUERY,
    SET_PAGE_NUMBER,
} = API_KEYS_MUTATIONS;

export class ApiKeysState {
    public cursor: ApiKeyCursor = new ApiKeyCursor();
    public page: ApiKeysPage = new ApiKeysPage();
    public selectedApiKeysIds: string[] = [];
}

/**
 * creates apiKeys module with all dependencies
 *
 * @param api - apiKeys api
 */
export function makeApiKeysModule(api: ApiKeysApi): StoreModule<ApiKeysState> {
    return {
        state: new ApiKeysState(),
        mutations: {
            [SET_PAGE](state: ApiKeysState, page: ApiKeysPage) {
                state.page = page;
                state.page.apiKeys = state.page.apiKeys.map(apiKey => {
                    if (state.selectedApiKeysIds.includes(apiKey.id)) {
                        apiKey.isSelected = true;
                    }

                    return apiKey;
                });
            },
            [SET_PAGE_NUMBER](state: ApiKeysState, pageNumber: number) {
                state.cursor.page = pageNumber;
            },
            [SET_SEARCH_QUERY](state: ApiKeysState, search: string) {
                state.cursor.search = search;
            },
            [CHANGE_SORT_ORDER](state: ApiKeysState, order: ApiKeyOrderBy) {
                state.cursor.order = order;
            },
            [CHANGE_SORT_ORDER_DIRECTION](state: ApiKeysState, direction: SortDirection) {
                state.cursor.orderDirection = direction;
            },
            [TOGGLE_SELECTION](state: ApiKeysState, apiKey: ApiKey) {
                if (!state.selectedApiKeysIds.includes(apiKey.id)) {
                    apiKey.isSelected = true;
                    state.selectedApiKeysIds.push(apiKey.id);

                    return;
                }

                apiKey.isSelected = false;
                state.selectedApiKeysIds = state.selectedApiKeysIds.filter(apiKeyId => {
                    return apiKey.id !== apiKeyId;
                });
            },
            [CLEAR_SELECTION](state: ApiKeysState) {
                state.selectedApiKeysIds = [];
                state.page.apiKeys = state.page.apiKeys.map((apiKey: ApiKey) => {
                    apiKey.isSelected = false;

                    return apiKey;
                });
            },
            [CLEAR](state: ApiKeysState) {
                state.cursor = new ApiKeyCursor();
                state.page = new ApiKeysPage();
                state.selectedApiKeysIds = [];
            },
        },
        actions: {
            fetchApiKeys: async function ({commit, rootGetters, state}, pageNumber: number): Promise<ApiKeysPage> {
                const projectId = rootGetters.selectedProject.id;
                commit(SET_PAGE_NUMBER, pageNumber);

                const apiKeysPage: ApiKeysPage = await api.get(projectId, state.cursor);
                commit(SET_PAGE, apiKeysPage);

                return apiKeysPage;
            },
            createApiKey: async function ({commit, rootGetters}: any, name: string): Promise<ApiKey> {
                const projectId = rootGetters.selectedProject.id;

                const apiKey = await api.create(projectId, name);

                return apiKey;
            },
            deleteApiKey: async function({state, commit}: any): Promise<void> {
                await api.delete(state.selectedApiKeysIds);

                commit(CLEAR_SELECTION);
            },
            setApiKeysSearchQuery: function ({commit}, search: string) {
                commit(SET_SEARCH_QUERY, search);
            },
            setApiKeysSortingBy: function ({commit}, order: ApiKeyOrderBy) {
                commit(CHANGE_SORT_ORDER, order);
            },
            setApiKeysSortingDirection: function ({commit}, direction: SortDirection) {
                commit(CHANGE_SORT_ORDER_DIRECTION, direction);
            },
            toggleApiKeySelection: function ({commit}, apiKey: ApiKey): void {
                commit(TOGGLE_SELECTION, apiKey);
            },
            clearApiKeySelection: function ({commit}): void {
                commit(CLEAR_SELECTION);
            },
            clearApiKeys: function ({commit}): void {
                commit(CLEAR);
                commit(CLEAR_SELECTION);
            },
        },
        getters: {
            selectedApiKeys: (state: ApiKeysState) => state.page.apiKeys.filter((key: ApiKey) => key.isSelected),
            apiKeys: function (state: ApiKeysState): ApiKey[] {
                return state.page.apiKeys;
            },
        },
    };
}
