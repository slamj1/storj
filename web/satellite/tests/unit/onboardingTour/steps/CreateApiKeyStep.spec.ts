// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import Vuex from 'vuex';

import CreateApiKeyStep from '@/components/onboardingTour/steps/CreateApiKeyStep.vue';

import { makeApiKeysModule } from '@/store/modules/apiKeys';
import { makeProjectsModule } from '@/store/modules/projects';
import { ApiKeysPage } from '@/types/apiKeys';
import { Project } from '@/types/projects';
import { NotificatorPlugin } from '@/utils/plugins/notificator';
import { SegmentioPlugin } from '@/utils/plugins/segment';
import { createLocalVue, mount } from '@vue/test-utils';

import { ApiKeysMock } from '../../mock/api/apiKeys';
import { ProjectsApiMock } from '../../mock/api/projects';

const localVue = createLocalVue();
const notificationPlugin = new NotificatorPlugin();
const segmentioPlugin = new SegmentioPlugin();
const projectsApi = new ProjectsApiMock();
const projectsModule = makeProjectsModule(projectsApi);
const apiKeysApi = new ApiKeysMock();
const apiKeysModule = makeApiKeysModule(apiKeysApi);
apiKeysApi.setMockApiKeysPage(new ApiKeysPage());
const project = new Project('id', 'projectName', 'projectDescription', 'test', 'testOwnerId', true);
projectsApi.setMockProjects([project]);

localVue.use(Vuex);
localVue.use(notificationPlugin);
localVue.use(segmentioPlugin);

const store = new Vuex.Store({ modules: { projectsModule, apiKeysModule }});

describe('CreateApiKeyStep.vue', () => {
    it('renders correctly', (): void => {
        const wrapper = mount(CreateApiKeyStep, {
            store,
            localVue,
        });

        expect(wrapper.findAll('.disabled').length).toBe(1);
        expect(wrapper).toMatchSnapshot();
    });

    it('create api key works correctly correctly', async (): Promise<void> => {
        const wrapper = mount(CreateApiKeyStep, {
            store,
            localVue,
        });

        await wrapper.vm.setApiKeyName('testName');
        await wrapper.vm.createApiKey();

        expect(wrapper.findAll('.disabled').length).toBe(0);
        expect(wrapper).toMatchSnapshot();
    });

    it('done click works correctly correctly', async (): Promise<void> => {
        const wrapper = mount(CreateApiKeyStep, {
            store,
            localVue,
        });

        await wrapper.find('.done-button').trigger('click');

        expect(wrapper.emitted()).toHaveProperty('setUploadDataState');
    });
});
