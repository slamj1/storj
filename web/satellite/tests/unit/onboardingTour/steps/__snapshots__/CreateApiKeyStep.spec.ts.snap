// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateApiKeyStep.vue create api key works correctly correctly 1`] = `
<div class="create-api-key-step">
  <h1 class="create-api-key-step__title">Create an API Key</h1>
  <p class="create-api-key-step__sub-title">
    API keys provide access to the project for creating buckets and uploading objects through the command line
    interface. This will be your first API key, and you can always create more keys later on.
  </p>
  <div class="create-api-key-step__container">
    <div class="create-api-key-step__container__title-area">
      <h2 class="create-api-key-step__container__title-area__title">Create API Key</h2>
      <!---->
    </div>
    <div class="input-container full-input">
      <div class="label-container">
        <!---->
        <h3 class="label-container__label">API Key Name</h3>
        <h3 class="label-container__label add-label"></h3>
        <!---->
      </div>
      <!---->
      <!----> <input id="API Key Name" placeholder="Enter API Key Name (i.e. Dan’s Key)" type="text" class="headered-input" style="height: 48px;">
    </div>
    <div class="create-api-key-step__container__copy-key-area">
      <div class="create-api-key-step__container__copy-key-area__header"><svg width="18" height="18" viewBox="0 0 18 18" fill="none" xmlns="http://www.w3.org/2000/svg" class="info-svg">
          <path fill-rule="evenodd" clip-rule="evenodd" d="M9 -7.86805e-07C4.02944 -1.22135e-06 1.22135e-06 4.02944 7.86805e-07 9C3.52265e-07 13.9706 4.02944 18 9 18C13.9706 18 18 13.9706 18 9C18 4.02944 13.9706 -3.52265e-07 9 -7.86805e-07ZM10 13C10 13.5523 9.55229 14 9 14C8.44772 14 8 13.5523 8 13C8 12.4477 8.44772 12 9 12C9.55229 12 10 12.4477 10 13ZM10 9C10 9.55228 9.55228 10 9 10C8.44771 10 8 9.55228 8 9L8 5C8 4.44771 8.44772 4 9 4C9.55229 4 10 4.44772 10 5L10 9Z" fill="white"></path>
        </svg> <span class="create-api-key-step__container__copy-key-area__header__title">
                    API Keys only appear here once. Copy and paste this key to your preferred method of storing secrets.
                </span></div>
      <div class="create-api-key-step__container__copy-key-area__key-container"><span class="create-api-key-step__container__copy-key-area__key-container__key">testKey</span>
        <div class="create-api-key-step__container__copy-key-area__key-container__copy-button">
          <div class="container blue-white" style="width: 81px; height: 40px;"><span class="label">Copy</span></div>
        </div>
      </div>
    </div>
    <p class="create-api-key-step__container__info">
      We don’t record your API Keys, which are only displayed once when generated. If you loose this
      key, it cannot be recovered – but you can always create new API Keys when needed.
    </p>
    <!---->
  </div>
  <div class="done-button container" style="width: 156px; height: 48px;"><span class="label">Done</span></div>
</div>
`;

exports[`CreateApiKeyStep.vue renders correctly 1`] = `
<div class="create-api-key-step">
  <h1 class="create-api-key-step__title">Create an API Key</h1>
  <p class="create-api-key-step__sub-title">
    API keys provide access to the project for creating buckets and uploading objects through the command line
    interface. This will be your first API key, and you can always create more keys later on.
  </p>
  <div class="create-api-key-step__container">
    <div class="create-api-key-step__container__title-area">
      <h2 class="create-api-key-step__container__title-area__title">Create API Key</h2>
      <!---->
    </div>
    <div class="input-container full-input">
      <div class="label-container">
        <!---->
        <h3 class="label-container__label">API Key Name</h3>
        <h3 class="label-container__label add-label"></h3>
        <!---->
      </div>
      <!---->
      <!----> <input id="API Key Name" placeholder="Enter API Key Name (i.e. Dan’s Key)" type="text" class="headered-input" style="height: 48px;">
    </div>
    <div class="create-api-key-step__container__create-key-area">
      <div class="generate-button container blue-white" style="width: 100%; height: 40px;"><span class="label">Generate API Key</span></div>
    </div>
    <!---->
    <!---->
  </div>
  <div class="done-button container disabled" style="width: 156px; height: 48px;"><span class="label">Done</span></div>
</div>
`;
